File: impls/cs/readline.cs
using System;
using Mono.Terminal; // LineEditor (getline.cs)

namespace Mal {
    public class readline {
        public enum Mode { Terminal, Raw };
        public static Mode mode = Mode.Terminal;

        static LineEditor lineedit = null;

        public static string Readline(string prompt) {
            if (mode == Mode.Terminal) {
                if (lineedit == null) {
                    lineedit = new LineEditor("Mal");
                }
                return lineedit.Edit(prompt, "");
            } else {
                Console.Write(prompt);
                Console.Out.Flush();
                return Console.ReadLine();
            }
        }
    }
}

File: impls/vb/readline.vb
Imports System
Imports Mono.Terminal ' LineEditor (getline.cs)

Namespace Mal
    Public Class readline
        Enum Modes
            Terminal
            Raw
        End Enum

        Public Shared mode As Modes = Modes.Terminal

        Shared lineedit As LineEditor = Nothing

        Public Shared Sub SetMode(new_mode As Modes)
            mode = new_mode
        End Sub

        Public Shared Function Readline(prompt As String) As String
            If mode = Modes.Terminal Then
                If lineedit Is Nothing Then
                    lineedit = New LineEditor("Mal")
                End If
                return lineedit.Edit(prompt, "")
            Else
                Console.Write(prompt)
                Console.Out.Flush()
                return Console.ReadLine()
            End If
        End Function
    End Class
End Namespace
